firedrake/dmplex.pyx

```diff
@@ -636,182 +636,83 @@ def mark_entity_classes(PETSc.DM plex):
-
-    # Mark the closures of non_core cells as non_core
-    ncells = plex.getStratumSize("op2_non_core", 1)
-    cell_is = plex.getStratumIS("op2_non_core", 1)
-    CHKERR(ISGetIndices(cell_is.iset, &cells))
-    for c in range(ncells):
-        if not (cStart <= cells[c] < cEnd):
-            continue
-
-        CHKERR(DMPlexGetTransitiveClosure(plex.dm, cells[c],
+    CHKERR(DMLabelCreateIndex(lbl_ghost, pStart, pEnd))
+    # If any entity in closure(cell) is in the halo, then all those
+    # entities in closure(cell) that are not in the halo are owned,
+    # but not core.
+    for c in range(cStart, cEnd):
+        CHKERR(DMPlexGetTransitiveClosure(plex.dm, c,
                                           PETSC_TRUE,
                                           &nclosure,
                                           &closure))
+        is_owned = PETSC_FALSE
         for ci in range(nclosure):
             p = closure[2*ci]
-            CHKERR(DMLabelHasPoint(lbl_exec, p, &has_point))
-            if not has_point:
-                CHKERR(DMLabelSetValue(lbl_non_core, p, 1))
-
-    # Mark all remaining points as core
-    pStart, pEnd = plex.getChart()
-    CHKERR(DMLabelCreateIndex(lbl_non_core, pStart, pEnd))
-    for p in range(pStart, pEnd):
-        CHKERR(DMLabelHasPoint(lbl_non_core, p, &is_non_core))
-        CHKERR(DMLabelHasPoint(lbl_exec, p, &is_exec))
-        if not is_exec and not is_non_core:
-            CHKERR(DMLabelSetValue(lbl_core, p, 1))
-    CHKERR(DMLabelDestroyIndex(lbl_non_core))
-
-    # Halo facets that only touch halo vertices and halo cells need to
-    # be marked as non-exec.
-    nfacets = plex.getStratumSize("op2_exec_halo", 1)
-    facet_is = plex.getStratumIS("op2_exec_halo", 1)
-    CHKERR(ISGetIndices(facet_is.iset, &facets))
-    for f in range(nfacets):
-        if not (fStart <= facets[f] < fEnd):
-            continue
-
-        non_exec = PETSC_TRUE
-        # Check for halo vertices
-        CHKERR(DMPlexGetTransitiveClosure(plex.dm, facets[f],
-                                          PETSC_TRUE,
-                                          &nclosure,
-                                          &closure))
-        for ci in range(nclosure):
-            if vStart <= closure[2*ci] < vEnd:
-                CHKERR(DMLabelHasPoint(lbl_exec, closure[2*ci], &has_point))
-                if not has_point:
-                    # Touches a non-halo vertex, needs to be executed
-                    # over.
-                    non_exec = PETSC_FALSE
-        if non_exec:
-            # If we still think we're non-exec, check for halo cells
-            CHKERR(DMPlexGetTransitiveClosure(plex.dm, facets[f],
-                                              PETSC_FALSE,
-                                              &nclosure,
-                                              &closure))
+            CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
+            if is_ghost:
+                is_owned = PETSC_TRUE
```

Comment:
So a point in the closure is a ghost, so therefore all the points in the closure must be either owned or ghost.